### DISCUSSION TITLE
Skip a state based on a condition

### DISCUSSION THREAD

**Discussion by vsrkmovva** [statelyai/xstate#1972] [Ideas]
Looking for any existing features for the following requirements

A way to skip a state based on a condition and automatically continue to next step using guards.
I am using json-rule-engine to evaluate the condition, where the facts are dynamic.

A way to list the completed states.

A way to see the pending states.

A way to rerun with skipping already completed states.

**Comment by ezhikov** [comment]
You can easily skip state with [`always`](https://xstate.js.org/docs/guides/transitions.html#eventless-always-transitions) property. Just attach a guard and it will not transition if condition returned false.

And I'm not sure about the list of pending and completed states. What do you mean by that?

As of skipping completed, you can model each state as hierarchy of `pending` and `completed` states. With history you can reenter last active state and, there you can again have `always` transition to immediately transition to another state.

**Reply by vsrkmovva** [reply]
I have to define a dynamic flow of UI dialogs based on a different different Workplans. It is all dynamic. Not fixed json. Each plan will have different steps/states. Looking to have a generic solution 

1. I should be able to run the UI wizard based on this state tree configuration
2. Assume for one step if user does not have the permission at any state(UI data entry), we should be able to skip it. i.e, we should be able skip and continue to next state(next UI widget). No need to to show that.
3. I should be able to know what are all the steps that did not skip. Able to track it with Context.
4. When another user who have access to those skipped UI state started, I should be able to rerun the automated flow again. As context have completed state info, I can check this in condition.
5. At any point of time, I need to show the progress of the flow, i.e, the complete and incomplete states



Here is the sample

`{
  "key": "light",
  "initial": "Step1",
  "states": {
    "Step1": {
      "on": {
        "FINISH": "Step2"
      },
      "initial": "PERM1",
      "states": {
        "PERM1": {
          "on": {
            "FINISH": "PERM2"
          }
        },
        "PERM2": {
          "on": {
            "FINISH": "PERM3"
          }
        },
        "PERM3": {}
      }
    },
    "Step2": {
      "on": {
        "FINISH": "Step3"
      },
      "initial": "SERVICE-PERM",
      "states": {
        "SERVICE-PERM": {
          "on": {
            "FINISH": "TREATMENT-WORKPLAN"
          }
        },
        "TREATMENT-WORKPLAN": {
          "on": {
            "FINISH": "WORKPLAN"
          },
          "initial": "Step1",
          "states": {
            "Step1": {
              "on": {
                "FINISH": "Step2"
              }
            },
            "Step2": {
              "on": {
                "FINISH": "Step3"
              }
            },
            "Step3": {}
          }
        },
        "WORKPLAN": {
          "initial": "Step1",
          "states": {
            "Step1": {
              "on": {
                "FINISH": "Step2"
              }
            },
            "Step2": {
              "on": {
                "FINISH": "Step3"
              }
            },
            "Step3": {}
          }
        }
      }
    },
    "Step3": {
      "on": {
        "FINISH": "ClosingPlan"
      },
      "initial": "PERM7",
      "states": {
        "PERM7": {
          "on": {
            "FINISH": "PERM8"
          }
        },
        "PERM8": {
          "on": {
            "FINISH": "PERM9"
          }
        },
        "PERM9": {}
      }
    },
    "ClosingPlan": {
      "on": {
        "FINISH": "Step1"
      },
      "initial": "PERM10",
      "states": {
        "PERM10": {
          "on": {
            "FINISH": "PERM11"
          }
        },
        "PERM11": {
          "on": {
            "FINISH": "PERM12"
          }
        },
        "PERM12": {}
      }
    }
  }
}`

**Reply by vsrkmovva** [reply]
![image](https://user-images.githubusercontent.com/40064833/110197586-894f9e00-7e72-11eb-867a-d130f3af5d0f.png)

**Reply by vsrkmovva** [reply]
Able to achieve straight forward flow, with all my events as finish.

Based on current state name, I am able to render angular UI components.

On finish click of each UI dialog, sending 'FINISH' event to interpreter.

It is changing the state and displaying the UI.

How to achieve  **skip the state** based on a condition (may be user do not have permission to fill that UI).
the only event I have is 'FINISH'. 

Machine should halt in a state wherever the condition met, i.e., user has permission.  One at a time.

My condition is also single. Have permission or not.

My machine is not static. It is dynamic, so not able to write a static machine config.

Above JSON is generated from DB. An admin can do any plan with any order. It is not fixed.

Like this there can be number of workflows.

**Reply by ezhikov** [reply]
I made a sandbox to show how you can track completed and check permissions.  https://codesandbox.io/s/frosty-grass-lofte?file=/src/index.js

It is way simpler than yours, just to show idea. You can put substeps into "pending" state of steps and  modify action that saves completed so it would give you more information
